pandas/core/frame.py

```diff
@@ -10461,54 +10460,44 @@ def _get_data() -> DataFrame:
                 data = self._get_bool_data()
             return data
 
-        if numeric_only or axis == 0:
-            # For numeric_only non-None and axis non-None, we know
-            #  which blocks to use and no try/except is needed.
-            #  For numeric_only=None only the case with axis==0 and no object
-            #  dtypes are unambiguous can be handled with BlockManager.reduce
-            # Case with EAs see GH#35881
-            df = self
-            if numeric_only:
-                df = _get_data()
-            if axis == 1:
-                df = df.T
-                axis = 0
-
-            # After possibly _get_data and transposing, we are now in the
-            #  simple case where we can use BlockManager.reduce
-            res = df._mgr.reduce(blk_func)
-            out = df._constructor(res).iloc[0]
-            if out_dtype is not None:
-                out = out.astype(out_dtype)
-            if axis == 0 and len(self) == 0 and name in ["sum", "prod"]:
-                # Even if we are object dtype, follow numpy and return
-                #  float64, see test_apply_funcs_over_empty
-                out = out.astype(np.float64)
-
-            return out
-
-        assert not numeric_only and axis in (1, None)
-
-        data = self
-        values = data.values
-        result = func(values)
-
-        if hasattr(result, "dtype"):
-            if filter_type == "bool" and notna(result).all():
-                result = result.astype(np.bool_)
-            elif filter_type is None and is_object_dtype(result.dtype):
-                try:
-                    result = result.astype(np.float64)
-                except (ValueError, TypeError):
-                    # try to coerce to the original dtypes item by item if we can
-                    pass
-
+        # Case with EAs see GH#35881
+        df = self
+        if numeric_only:
+            df = _get_data()
         if axis is None:
-            return result
+            return func(df.values)
```

Comment:
hmm .values can be expensive, might be better to reduce twice?  (... which can also be expensive.  darn).  Is punting on this viable?

Reply:
Using `.values` is the current behavior on main if `numeric_only` is False and `axis` is not 0.

https://github.com/pandas-dev/pandas/blob/c7fa61113b4cf09581e8f31f6053f2e64b83a9fc/pandas/core/frame.py#L10494-L10496

In fact, main is currently broken when `numeric_only` is False and `axis` is None:

```
df = pd.DataFrame({'a': [1, 1, 2], 'b': [3, 4, 5], 'c': list('xyz')})
result = df.mean(axis=None, numeric_only=True)
print(result)
# a    1.333333
# b    4.000000
# dtype: float64

result2 = df[['a', 'b']].mean(axis=None)
print(result2)
# 2.6666666666666665
```

Assuming this is okay for now, I can add a test for when `numeric_only` is False and `axis` is None.

Reply:
@jbrockmendel - friendly ping